golang/cosmos/daemon/cmd/root.go

```diff
@@ -144,10 +170,27 @@ func initRootCmd(sender vm.Sender, rootCmd *cobra.Command, encodingConfig params
 		snapshot.Cmd(ac.newSnapshotsApp),
 	)
 
-	server.AddCommands(rootCmd, gaia.DefaultNodeHome, ac.newApp, ac.appExport, addModuleInitFlags)
+	server.AddCommands(rootCmd, gaia.DefaultNodeHome, ac.newApp, ac.appExport, addStartFlags)
 
 	for _, command := range rootCmd.Commands() {
 		switch command.Name() {
+		case "start":
+			var preRunE cobraRunE = func(cmd *cobra.Command, _ []string) error {
+				// Consume and validate config.
+				viper := server.GetServerContextFromCmd(cmd).Viper
+				baseConfig, err := serverconfig.GetConfig(viper)
+				if err != nil {
+					return err
+				}
+				if err = baseConfig.ValidateBasic(); err != nil {
+					return err
+				}
+				if _, err = swingset.SwingsetConfigFromViper(viper); err != nil {
+					return err
+				}
+				return nil
+			}
+			appendToPreRunE(command, preRunE)
```

Comment:
I'm surprised there isn't a better way to hook to validate custom config. Why not parse / validate the config in the `appCreator` ? We already cast `appOpts` to the `Viper` instance anyway. I would also expect the swingset config to be parsed for other commands, like snapshot restores, since that involves cosmic-swingset as well.

Reply:
`appCreator` would be great, but doesn't have an error return so we're limited to `panic` and `return nil`. But this pattern can also be applied to other commands if need be.

Reply:
Yeah it's really weird that `appCreator` doesn't get to return an error. Especially because the callers have an error returning path. It's unfortunate that it forces us to further intercept commands.

I'm wondering if it wouldn't be better to call this in our `PersistentPreRunE`, right after calling `server.InterceptConfigsPreRunHandler`, since that's what finishes all the config stuff, and looking at the internals of cobra, is called immediately before the intercepted `PreRunE` here.

Reply:
PersistentPreRunE also applies to client commands like `agd query`, for which it would be inappropriate to validate server config.

Reply:
Good call!